### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by jmkitavi**

### CODE REVIEW

File: app/config.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import os
+basedir = os.path.abspath(os.path.dirname(__file__))
+
+
+class Config(object):
+    DEBUG = True
+    TESTING = False
+    CSRF_ENABLED = True
+    SECRET_KEY =os.getenv('SECRET_KEY')
+    SQLALCHEMY_DATABASE_URI = 'postgresql+psycopg2://postgres@localhost/bucketlist'
```
Comment (nyambati) [inline app/config.py]: Use environment variable so as to make the URL dynamic when the app is exposed to different environment.

File: app/config.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    TESTING = True
+    # sqlite database for testing
+    SQLALCHEMY_DATABASE_URI = 'sqlite://'
+
+configuration = {
+    'staging': Staging,
+    'testing': Testing,
+    'development': Development,
+    'default': Config,
+    'SECRET_KEY': "SECRET_KEY"
```
Comment (nyambati) [inline app/config.py]: The secret key should be passed as environment variable for security resoans

File: app/views.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+    def post(self):
+        args = self.reqparse.parse_args()
+        username = args['username']
+        password = args['password']
+
+        # checking if parameters are null
+        if password == "" or username == "":
+            return {'error': "Username or Password can't be empty"}, 400
+            #  status code - Bad request
```
Comment (neddoz) [inline app/views.py]: You could have the comment come before the code

File: app/views.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    'date_created': fields.DateTime,
+    'date_modified': fields.DateTime,
+    'created_by': fields.Integer,
+    # 'items': fields.Nested(format_item),
+}
+
+
+class BucketListAPI(Resource):
+    decorators = [auth.login_required]
+    """ One bucket list
```
Comment (neddoz) [inline app/views.py]: same here. The docstring should come immediately after the class declaration

File: app/views.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        self.reqparse.add_argument(
+            'description', type=str, default="", location='json')
+
+        args = self.reqparse.parse_args()
+        title = args['title']
+        description = args['description']
+        user_id = g.user.user_id
+
+        # check if title is null
+        if title == "":
```
Comment (neddoz) [inline app/views.py:191]: Repetition. You have already validated the title can't be empty this using reqparse.

File: app/views.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+        bucketlist = BucketList(
+            title=title, description=description, created_by=user_id)
+        db.session.add(bucketlist)
+        db.session.commit()
+        return {'message': '{} - bucketlist has been added succesfully by {}'.format(
+            title, g.user.username)}, 201
+        # status code - created new resource
+
+    # add pagination
```
Comment (neddoz) [inline app/views.py]: Were these left here intentionally or you forgot to do some clean up.

File: app/views.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+                                   help="Item name can't be blank", location='json')
+        self.reqparse.add_argument('status', type=bool, location='json')
+
+        args = self.reqparse.parse_args()
+        item_name = args['item_name']
+        status = args['status']
+        user_id = g.user.user_id
+
+        # check if title is null
+        if item_name == "":
```
Comment (neddoz) [inline app/views.py:288]: Repetition. Already validated by reqparse

File: app/views.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        if BucketList.query.filter_by(bucketlist_id=bucketlist_id).first() is None:
+            return {'error': 'BucketList ID {} not found.'.format(bucketlist_id)}, 404
+            # status code - Not found
+
+        # check if item name exists in bucketlist
+        if BucketListItems.query.filter_by(item_name=item_name,
+                                           bucketlist_id=bucketlist_id).first() is not None:
+            return {'error': 'BucketList {} already has item with name {}'.format(
+                bucketlist_id, item_name)}, 202
+            #  status code - request accepted but not processed
```
Comment (neddoz) [inline app/views.py]: Again. The inline comment should come just before the code

File: app/views.py (hunk 5 above)
Comment (jmkitavi) [reply app/views.py:191]: @neddoz I realized that reqparse only checks if the parameter has been passed in the request, but doesn't check if its actually an empty string.

Comment (neddoz) [reply app/views.py:191]: @sirjmkitavi ooh. i see.

File: app/views.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+    'description': fields.String,
+    'date_created': fields.DateTime,
+    'date_modified': fields.DateTime,
+    'created_by': fields.Integer,
+    # 'items': fields.Nested(format_item),
+}
+
+
+class BucketListAPI(Resource):
+    """ One bucket list
```
Comment (mrmuli) [inline app/views.py:113]: What's up with your comments and docstrings?

**Review by mrmuli — COMMENTED** [review]
Your inline comments aren't pythonic, ensure you follow PEP8 commenting standards
Try as much as possible and reduce the load in views.py, that is separate your work. Let auth be seperate from request methods
Also ensure you follow REST standards and use serializers for your requests

**Pull request merged**